### PR TITLE
Ser 160 modify registration user flow

### PR DESCRIPTION
**Pull request by shejasabin**
SER-160: Modified Kenya registration service to include a group leader interest question.
SER-162: modify confirmation message if the FO number is not found
SER-4: save all the below info in a dataTable

user flow:
- A client logs in to the system using their account number. you can use: **24450523**
-  Chooses option 3 Enrollment
-  Chooses 0(register new client)
-  Enter the national ID of the client being registered(use 98765432 or 12332112)
- Confirm their national Id
- Enter their first name
- Enter their last name
- Enter their phone number(can use **0712301351**)
- Confirm phone number
-  **Answer if they would like to be a GL or not** (new)
- Receive after containing the account number of the farmer
- We save the above info in a dataTable 'dev_lr_2021_client_table'

For returning clients that enroll, the only change is the message we send if there is no fo Phone retrieved from roster and saving the info.

Steps: 
-log in with an account number: can use **24142342**
-choose option 3
-Enter the client's account number: can use **24450523**

link: https://telerivet.com/p/0c6396c9/service/SVb47209de1fdd2cf0/edit

### CODE REVIEW

File: client-registration/group-leader-question-handler/test.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    it('should not call onGroupLeaderQuestion if input is not 1 or 2', () => {
+        groupLeaderQuestionHandler('7');
+        expect(onGroupLeaderQuestion).not.toHaveBeenCalled();
+    });
+    it('should set groupLeader state variable to true and call onGroupLeaderQuestion if input is 1', () => {
+        groupLeaderQuestionHandler('1');
+        expect(state.vars.groupLeader).toBeTruthy();
+        expect(onGroupLeaderQuestion).toHaveBeenCalled();
+    });
+    it('should set groupLeader state variable to true and call onGroupLeaderQuesion if the input is 2', () => {
```
Comment (nyabongo) [inline client-registration/group-leader-question-handler/test.js:25]: I think this test is mislabelled

File: client-enrollment/clientEnrollment.js [hunk 2]
```diff
@@ -25,18 +26,31 @@ module.exports = {
...
                 var foInfo = getFOInfo(client.DistrictId,client.SiteId,state.vars.enr_lang);
-                console.log(foInfo);
-                if(foInfo){
-                    var message = translate('registration_message' , {'$phone': foInfo.phoneNumber}, state.vars.enr_lang);
-                    project.sendMessage({
-                        content: message, 
-                        to_number: contact.phone_number
-                    });
+                var message;
+                if(foInfo == null || (foInfo.phoneNumber == null || undefined)){
```
Comment (Bahatiroben) [inline client-enrollment/clientEnrollment.js:30]: it seems that statement (foInfo.phoneNumber == null || undefined)  will always return undefined if the first condition is false. I reckon (foInfo.phoneNumber == null || foInfo.phoneNumber == undefined) would be the correct statement

File: client-registration/clientRegistration.js [hunk 3]
```diff
@@ -53,23 +59,49 @@ module.exports = {
                 console.log('client Data ****************' + clientData.AccountNumber);
                 var getFOInfo = require('../Roster-endpoints/Fo-info/getFoInfo');
                 var foInfo = getFOInfo(clientData.DistrictId,clientData.SiteId,state.vars.reg_lang);
-                var message = translate('reg_complete_message' , {'$ACCOUNT_NUMBER': clientData.AccountNumber,'$FOphone': foInfo.phone}, state.vars.reg_lang);
+                var message;
+                if((foInfo == null) || (foInfo.phoneNumber == null || undefined)){
```
Comment (Bahatiroben) [inline client-registration/clientRegistration.js:63]: it seems that statement (foInfo.phoneNumber == null || undefined)  will always return undefined if the first condition is false. I reckon (foInfo.phoneNumber == null || foInfo.phoneNumber == undefined) would be the correct statement

File: client-enrollment/test.js [hunk 4]
```diff
@@ -57,6 +61,51 @@ describe('clientRegistration', () => {
...
+            });
+        });
+        it('should send a message without FO contact if loan is fully paid and foinfo is null',()=>{
+            var {client}  = require('./test-client-data'); 
+            roster.getClient = jest.fn().mockImplementationOnce(() => {return client ;});
+            getFOInfo.mockImplementationOnce(() => {return null;});
+            clientEnrollment.start(account, country, enr_lang);
+            expect(project.sendMessage).toHaveBeenCalledWith({
+                content: 'Thank you for expressing your interest to enroll with OAF. Your FO will reach out to you to add inputs to your order.', 
+                to_number: contact.phone_number
```
Comment (Bahatiroben) [inline client-enrollment/test.js:85]: I guess expecting a return value by a matching string literal is more efficient that expecting a value from a variable. something like *to_number: '07847358323'* would be more efficient.

The same for all test cases that expect by a variable.

File: client-enrollment/clientEnrollment.js (hunk 2 above)
Comment (shejasabin) [reply client-enrollment/clientEnrollment.js]: ooooh, yeah. Fixed

File: client-registration/group-leader-question-handler/test.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        expect(groupLeaderQuestionHandler).toBeInstanceOf(Function);
+    });
+    it('should not call onGroupLeaderQuestion if input is not 1 or 2', () => {
+        groupLeaderQuestionHandler('7');
+        expect(onGroupLeaderQuestion).not.toHaveBeenCalled();
+    });
+    it('should set groupLeader state variable to true and call onGroupLeaderQuestion if input is 1', () => {
+        groupLeaderQuestionHandler('1');
+        expect(state.vars.groupLeader).toBeTruthy();
+        expect(onGroupLeaderQuestion).toHaveBeenCalled();
```
Comment (Bahatiroben) [inline client-registration/group-leader-question-handler/test.js:23]: in the call of *onGroupLeaderQuestion* it receives a parameter. so, toHaveBeenCalledWith('something') would be better if it gets called with a parameter.

File: client-enrollment/test.js (hunk 4 above)
Comment (shejasabin) [reply client-enrollment/test.js:85]: I'm confused. Are we using strings or variables? I think we should stick to one. @nyabongo?

File: client-registration/group-leader-question-handler/test.js (hunk 1 above)
Comment (shejasabin) [reply client-registration/group-leader-question-handler/test.js:25]: oooh, yeah. fixing it

File: client-registration/group-leader-question-handler/test.js (hunk 5 above)
Comment (shejasabin) [reply client-registration/group-leader-question-handler/test.js:23]: oooh, yeah, fixing it. sorry about that

**Comment by Bahatiroben** [comment]
> SER-160: Modified Kenya registration service to include a group leader interest question.
> SER-162: modify confirmation message if the FO number is not found
> SER-4: save all the below info in a dataTable
> 
> user flow:
> 
> * A client logs in to the system using their account number. you can use: **24450523**
> * Chooses option 3 Enrollment
> * Chooses 0(register new client)
> * Enter the national ID of the client being registered(use 98765432 or 12332112)
> * Confirm their national Id
> * Enter their first name
> * Enter their last name
> * Enter their phone number(can use **0712301351**)
> * Confirm phone number
> * **Answer if they would like to be a GL or not** (new)
> * Receive after containing the account number of the farmer
> * We save the above info in a dataTable 'dev_lr_2021_client_table'
> 
> For returning clients that enroll, the only change is the message we send if there is no fo Phone retrieved from roster and saving the info.
> 
> Steps:
> -log in with an account number: can use **24142342**
> -choose option 3
> -Enter the client's account number: can use **24450523**

and a link to a telerivet service so that we can test the implementation.

**Pull request merged**